### ISSUE TITLE
Explore: should moduloze also support converting "globals-only" (instead of just CJS)

### ISSUE THREAD

**Issue by getify** [getify/moduloze#8]
https://twitter.com/LeaVerou/status/1285320583408955401?s=20

**Comment by MartinKolarik** [comment]
I might be missing something obvious, but two questions:

1. Shouldn't `bar` be exported as well since it's in the global scope?
2. I briefly checked how moduloze currently works with other formats and I noticed there's a required config option `dep-map`. I guess it's needed to map `require()` to UMD globals but I see it's somehow used in ESM processing as well. With an empty map, both ESM and UMD outputs look like I'd expect though. Can you explain briefly why the map is needed to make sure I understand the current approach for UMD/ESM?

**Comment by getify** [comment]
> Shouldn't bar be exported as well since it's in the global scope? Or do you expect the user to specify which (only one) global is exported?

Yep, it probably should be exported as well. Just overlooked it. Though, that does bring up pesky questions...

1. basically, all global declarations in such a file could be assumed to be needed, so they can all be named exports easily for both ESM and the CJS side of UMD. But what about multiple named exports for the browser globals and AMD sides of UMD? it doesn't really support that.

2. if there's only one global, should it also be a default export (ESM)? should it be the "main export" of the UMD form? would this be INSTEAD OF, or IN ADDITION TO, the named export?

3. but if there's multiple globals, do we do no default exports, or just pick the first one?

**Comment by MartinKolarik** [comment]
The way I'm thinking here is basically `window` being equivalent to `module.exports` for modules using globals (although you don't need to explicitly assign stuff to `window` for it to be global).

So, for ESM, I'd expect everything to be in named exports and accessed like this:

```js
// module user file
import { MyModule, bar } from 'm';
MyModule.foo();
bar();
```

For UMD-global:
```js
const { MyModule, bar } = m;
```

For UMD-AMD:

```js
define(['m'], function ({ MyModule, bar }) {
  ...
});
```

I think that answers 1 and 2, and as for 3, I'd say no default export. It would be bad if a seemingly non-important change in the input (adding a new global), changed how the output can be consumed.

**Comment by getify** [comment]
> I think that answers 1 and 2

Eh... not exactly. You've implied a change to the way the UMD boilerplate works, where it's assuming a destructuring of a namespace as opposed to just getting a single value out.

But I can't really think of any way for moduloze to determine, reliably, that you definitely want the destructuring applied for one module's imports, and not for another's (the current status quo). For example:

```js
define(['m','p'],function DEF({ MyModule, bar},Whatever){
   // ..
});

It would have to somehow semantically analyze how a module was being used after import and make a guess. That seems intractably complex given that we're talking about a non-static-module-format in a non-static language.

**Comment by MartinKolarik** [comment]
Well, a namespace (object) *is* just a single value. The `define` piece of code including the destructuring would be written by the user of the "global module", wouldn't it? E.g.

```js
// moduloze output
(function UMD(...){
   // ..
})(
   "m",
   window,
   {},
   function DEF(){
      var MyModule = {
         someData: 42,
         foo() { console.log(this.someData); }
         bar,
      };

      function bar() {
         console.log(this.someData * 2);
      }

      return { MyModule, bar };
   }
);
```

```js
// somewhere in the app using the generated file; written by the user of "m"
define(['m','whatEverOtherDeps'],function DEF({ MyModule, bar}, Whatever){
   // ..
});
```

Moduloze itself would only need to care about how to write that `define` if we considered "global dependencies", i.e, you try to convert ModuleA, which itself accesses ModuleB as a global - not sure if you intended, or thought about supporting that as well, but if we only consider "exporting" of globals, I don't see any problem.

**Comment by getify** [comment]
I was assuming that these two snippets from your previous message were implying what the **UMD boilerplate itself** would be doing, which has destructuring in it:

```js
const { MyModule, bar } = m;
```

```js
define(['m'], function ({ MyModule, bar }) {
  ...
});
```

Did I misunderstand your intent there?

**Comment by MartinKolarik** [comment]
Yes, those were meant to show how you would *consume* the processed file. `m` in both is whatever moduloze generated.

**Comment by getify** [comment]
OK, but I thought those you were saying those lines themselves were actually *in the UMD boilerplate*, seems like I misunderstood.

Let me illustrate the problem this way:

If you take the current snippet, and load that JS file with a `<script>` tag in a page, you end up with two globals, `MyModule` and `bar`. But if you "built" that file with this hypothetical mode of moduloze into UMD format, and loaded *that* UMD format into the browser with a `<script>` tag, you have just one global named `m` or `Mz843242394` or whatever.

See this line of the UMD boilerplate:

https://github.com/getify/moduloze/blob/0e20280ae0fadf950a2c1fbe97ee602e9cac19df/src/umd-template.js#L12

The `context[name] = ..` statement just creates a single global variable of the `name` name, so since there's no destructuring there, it won't be creating the multiple global declarations (`MyModule` and `bar`).

Typically, the "global mode" (via `<script>` load) of a UMD wrapper is supposed to produce the same sort of effect as if the file wasn't wrapped in the UMD wrapper. If the .js file has a single global declaration (aka "export"), that could still be the case. But in the case where your .js file is declaring multiple globals, the typical UMD boilerplate doesn't have a way to create all those globals from its single "export" assignment.

**Comment by MartinKolarik** [comment]
> The context[name] = .. statement just creates a single global variable of the name name, so since there's no destructuring there, it won't be creating the multiple global declarations (MyModule and bar).

I actually saw this as one of the goals of this feature. My assumption was that you have a file with many globals and you want to "wrap it up" so that it doesn't pollute the global namespace and is still usable via all the regular methods.

So when being consumed as a UMD global, the global would be a single object, which has all the things that were originally global. That would, of course, mean that when using the converted module, instead of referencing `MyModule` directly, you reference it as `m.MyModule` or include the destructuring line at the top of your file.

**Comment by getify** [comment]
> I actually saw this as one of the goals of this feature. 

It *could* be, it's just surprising to me because it's not how UMD's global behavior is typically construed. I'm just **very used to** being able to load a UMD via a `<script>` and get basically the same result as if I loaded the source itself. I may be projecting those assumptions onto others, but I was just thinking it would surprise others too.

**Comment by MartinKolarik** [comment]
> That makes sense (possibly) from the author's perspective, but I don't think users normally think that way. I think they're used to just consuming a dist file (from a CDN) in whatever their typical module-context is, and it "just works"... so for those who have been used to loading UMDs with <script>, it might surprise them. Maybe?

Here I think a question is in what scenarios you expect this functionality to be used:

1. By the author of a library, who, for whatever reason, wrote it using globals but wants to distribute it as a UMD. Here I don't see any issue at all because the users will consume it however the documentation says, without knowing the background of how the dist file was generated.

2. By a user of a library, which is only distributed via globals (and the user wants to use UMD instead). Here the user is interacting with moduloze directly and it's up to moduloze docs to explain the format of the output.

But even if this surprises someone, it should be an easy thing to notice and change the code accordingly.

Then again, there is an alternative - change the UMD wrapper so that for the global case, it reexports all variables under their original global names. I'm don't have a strong opinion here.
